### PR TITLE
Fixed everything Gábor Horváth commented on

### PR DESCRIPTION
**Pull request by Szelethus**
IMPORTANT NOTE: It looks like my previous commit broke the tests, making check-clang fail with 3 errors. Those are also fixed in this PR.

### CODE REVIEW

**Comment by Szelethus** [comment]
Gábor Horváth also mentioned running templight on most Sema test cases to see wether it produces a crash. That is still left to be done.

> I would bet the sema tests are full of tricky edge cases. So running templight on those tests would be a helpful exercise. I am only interested in assertion fails, so we do not need to check the output. One option to do so would be to add the -templight-dump option to the %clang_cc1 variable when running the tests. Note that the tests are likely to fail because the output will change, but if there are no crashes, it is fine.

**Comment by sabel83** [comment]
Did you check this as well?
https://reviews.llvm.org/D5767#inline-375405

File: templight_clang_patch.diff [hunk 1]
```diff
@@ -503,25 +502,26 @@ Index: lib/Sema/SemaTemplateInstantiate.cpp
...
 +    break;
    }
  
    llvm_unreachable("Invalid SynthesisKind!");
-@@ -235,6 +239,7 @@
+@@ -235,6 +240,7 @@
          !SemaRef.InstantiatingSpecializations
               .insert(std::make_pair(Inst.Entity->getCanonicalDecl(), Inst.Kind))
               .second;
 +    atTemplateBegin(SemaRef.TemplateInstCallbacks, SemaRef, Inst);
```
Comment (sabel83) [inline templight_clang_patch.diff]: Is this fixing a bug in the patch? (Does it fix something Gábor pointed out, or is it something different?)

Also: do any of the tests cover this? (Is there a test which breaks when this is removed?)

**Comment by Szelethus** [comment]
I did check the possibly unnecessary include, in fact I left a comment, but it is apparently "unsubmitted". In Sema.h, `TemplateInstantiationCallback` is only forward declared, so the include is necessary to provide a definition in the TU.

File: templight_clang_patch.diff (hunk 1 above)
Comment (Szelethus) [reply templight_clang_patch.diff]: Line 520? I never changed that particular line to the best of my knowledge.
During the summer I did a "simulated tdd", trying to recompile clang with none and gradually more and more changes from templight. I essentially tested whether any of the code is unnecessary. I'm not 100% sure I checked that particular line on its own, but I'm very much certain its needed.
